### ISSUE TITLE
Exporting Weights from Tensorflow Keras to C produces a different set of outputs, reduced accuracy in the model.

### ISSUE THREAD

**Issue by BeverlyCode** [tensorflow/tensorflow#58861]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Source

source

### Tensorflow Version

Any

### Custom Code

Yes

### OS Platform and Distribution

Linux

### Current Behaviour?

For some reason when I export my weights from a Tensorflow Keras model, be it a simple Sequential FNN, and then load them into C to perform forward passes I get different output results, sometimes similar and often just not the same at all.

This is how I export my weights:
```
    for layer in model.layers:
        if layer.get_weights() != []:
            numpy.savetxt(layer.name + ".csv", layer.get_weights()[0].flatten(), delimiter=",")
            numpy.savetxt(layer.name + "_bias.csv", layer.get_weights()[1].flatten(), delimiter=",")
```
and
```
f.write(", /* weight */ " + str(layer.get_weights()[0].flatten()[i]))
f.write(", /* bias */ " + str(layer.get_weights()[1].flatten()[i]))
```
The first method writes a long E value and the second writes a smaller float value to the most significant decimal place. Although that should not cause a loss in precision over the first method? right?

I'm not sure why the saved weights would give different outputs, even with a 1 layer FNN that has 1 output,  a very simple network, all the weights are saved to float32 in C and then processed with float math functions such as tanhf() etc depending on the layer output.

I must be missing some understanding of what Tensorflow or Keras is doing behind the scenes? Something it's default configuration is doing to my network/weights that I need to disable or reproduce? Are the arrays of weights in reverse or something? Do I need to iterate them from left to right and not right to left on the inputs? What am I missing?

Here is a project that demonstrates this issue:
https://github.com/jcwml/neural_unitvector

this is a good example project. weights trained in Tensorflow Keras with the same input data produces different outputs when processed in C.

In this project a dataset is exported from main.c to ensure the C program and Tensorflow Keras program both use data generated by the same random function, then Keras trains a set of weights for a simple FNN network to some number of hidden units and layers, and then those weights are exported and loaded into the C program where the forward pass is reconstructed in C code. There are 3 types of networks in this example that have had their forward passes reconstructed, each more complicated than the last, and yet, none of them produce the same outputs as Keras predict() had, it's not even a minor accuracy loss, it is substantially different, you can tell that the weights still represent what they where trained to but with just much worse prediction accuracy.

Another important fact to note is that when I train a network with multiple layers it would seem in many cases when I export the weights to C for a forward pass that the input data passed into the network eventually becomes 0's before it gets to the deeper layers, like vanishing gradients, this would imply that Tensorflow is using a different precision float format but the documentation says it uses float32 by default and the console printouts also claim to be using float32 in my application. My only thought is that maybe numpy is not saving them in the full float32 precision but when I looked into that apparently it does export full precision, and that exporting weights this way should be fine.

I am very confused.

**Comment by krrishdholakia** [comment]
Make sure that the weights are being loaded and processed correctly in C. You can do this by printing out the weights after loading them from the .csv files and checking them against the outputs from the model in Python. If they are not the same, then the weights are not being read in correctly. Check the data types being used in Python and C, and make sure they match. Additionally, make sure that the weights are being processed correctly in C, and that the same operations are being used that are used in the model in Python.

Btw used Clerkie (ai code debugging tool) - https://clerkie.co/, to help me brainstorm on approaches here (still reviewed/investigated feedback before sharing). Hope it helped!

<img width="1523" alt="Screen Shot 2022-12-13 at 12 28 31 PM" src="https://user-images.githubusercontent.com/17561003/207247697-9038576a-e10c-493b-8241-d173c8e314db.png">

**Comment by BeverlyCode** [comment]
Thank you [krrishdholakia](https://github.com/krrishdholakia) but that is not the problem, I have tested this with many simple FNN networks including one layer with one hidden unit straight to output which are very simple to implement and hard to get wrong, I have actually had this problem for a year or more and it has always puzzled me. It seems to be related to how Tensorflow is working behind the scenes, documentation claims it uses float32 by default, and so does the console of my python application when I check, and I also use float32 in C, so I am not sure where the discrepancy is coming from, that is why I am hoping someone more experienced in Tensorflow can shed some light on this problem.

**Comment by mohantym** [comment]
Hi @BeverlyCode !
Sorry for the late reply. Thanks for sharing your observation on difference in C++ upon loading the weights from Keras model trained in Python.

Would it be possible for you to create a toy model with [Tensorflow C++ ](https://www.tensorflow.org/install/lang_c) Documentation and post your results here to replicate the same inconsistency . 

You can post this query on [TF Forum](https://discuss.tensorflow.org/tag/none) too .

Thank you!

**Comment by BeverlyCode** [comment]
Hi so the problem is not to do with Tensorflow C++, it is that in Tensorflow Keras in Python that when I export the weights using the python code defined in my original issue that those same weights do not produce the same predictions (when I used them in a custom C program not using Tensorflow).

Weights trained in Tensorflow should work outside tensorflow using the name network topology, they don't and I don't know why. I would like to know why.

**Comment by mohantym** [comment]
Ok @BeverlyCode !
Thanks for the update. 

@sachinprasadhs !
Could you look at this issue.

Thank you!

**Comment by BeverlyCode** [comment]
Just a small update, from my research exporting the weights using both `numpy.savetxt(layer.name + ".csv", layer.get_weights()[0].flatten(), delimiter=",")` and `f.write("," + str(layer.get_weights()[0].flatten()[bc]))` on the latest Python versions and packages exports the weights to the exact decimal precision they are in.

But encase they didn't I also exported all the weight floats as bytes and then converted them back to floats in C to be extra-sure using this function on the export:
`def float_to_hex(f): return hex(struct.unpack('<I', struct.pack('<f', f))[0])` _(which gave the same weights)_

It would seem that in even very basic FNN's in Tensorflow, there is something that Tensorflow is doing in the forward-pass that is not obvious to me _(and seemingly between each layer if I had to guess)_, and I'm not sure how I would find-out what it is doing. I would assume that in Tensorflow the state of all things are disabled by default, such as layer regularisation, etc, unless explicitly turned on? Is it possible that Tensorflow actually turns some of this stuff on by default to improve default trained networks and that I just haven't realised this?

I'm clutching at straws here but Tensorflow is a large complicated SOTA framework and I'm not really sure where to start or how to go about debugging this.

But it would be very valuable to me to be able to take the weights of very simple FNN networks trained in Tensorflow Keras and then re-use them in a C program that re-creates the forward-pass.

When I export my model from a similar or equivalent FNN network made in Tensorflow Keras that I am also trying to export weights from in the exact same manner described above I get this output:
```
{
 "class_name": "Sequential",
 "config": {
  "name": "sequential",
  "layers": [
   {
    "class_name": "InputLayer",
    "config": {
     "batch_input_shape": [
      null,
      784
     ],
     "dtype": "float32",
     "sparse": false,
     "ragged": false,
     "name": "dense_input"
    }
   },
   {
    "class_name": "Dense",
    "config": {
     "name": "dense",
     "trainable": true,
     "dtype": "float32",
     "batch_input_shape": [
      null,
      784
     ],
     "units": 8,
     "activation": "tanh",
     "use_bias": true,
     "kernel_initializer": {
      "class_name": "GlorotUniform",
      "config": {
       "seed": null
      }
     },
     "bias_initializer": {
      "class_name": "Zeros",
      "config": {}
     },
     "kernel_regularizer": null,
     "bias_regularizer": null,
     "activity_regularizer": null,
     "kernel_constraint": null,
     "bias_constraint": null
    }
   },
   {
    "class_name": "Dense",
    "config": {
     "name": "dense_1",
     "trainable": true,
     "dtype": "float32",
     "units": 1,
     "activation": "sigmoid",
     "use_bias": true,
     "kernel_initializer": {
      "class_name": "GlorotUniform",
      "config": {
       "seed": null
      }
     },
     "bias_initializer": {
      "class_name": "Zeros",
      "config": {}
     },
     "kernel_regularizer": null,
     "bias_regularizer": null,
     "activity_regularizer": null,
     "kernel_constraint": null,
     "bias_constraint": null
    }
   }
  ]
 },
 "keras_version": "2.11.0",
 "backend": "tensorflow"
}
```
And this is odd because this output does seem to claim that it is just a regular basic FNN network, which my C code to reproduce the forward-pass of above in my prior responses should work just fine? It is odd.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
